Don't leak IPC connections in error/shutdown path

kinetiknz opened the pull request:
This fixes BMO 1607802, see https://bugzilla.mozilla.org/show_bug.cgi?id=1607802#c4 for bug analysis.

r? @ChunMinChang please

ChunMinChang commented on audioipc/src/lib.rs:
```diff
@@ -76,7 +80,8 @@ impl serde::Serialize for PlatformHandle {
     {
-        serializer.serialize_i64(self.0 as i64)
+        let h = self.0.borrow().unwrap_or(INVALID_HANDLE_VALUE);
+        serializer.serialize_i64(h as i64)
```
What happens if `INVALID_HANDLE_VALUE` is passed from one side to another? Would this _handle_ be checked before it's converted into a _file_?

kinetiknz replied:
I'll make this a panic instead, it's a bug if this happens.

ChunMinChang commented on audioipc/src/lib.rs:
```diff
@@ -140,21 +141,26 @@ impl PlatformHandle {
 
-    pub fn as_raw(self) -> PlatformHandleType {
-        self.0
+    pub unsafe fn as_raw(&self) -> PlatformHandleType {
```
Oh, is calling `take()` the reason the function names stay with `into*` or `as*`? Perhaps there are better names ...

kinetiknz replied:
I think `into` fits better since the source PlatformHandle becomes invalid.  Ideally PlatformHandle's `into_*` functions would take `self` instead, but that'll require significant changes to the API users.

kinetiknz replied:
I'll rename `as_raw` to `into_raw` and leave `into_file` as-is for now.

kinetiknz replied:
I'm using unsafe on `to_raw` to indicate the caller needs to take care not to leak the raw handles.

kinetiknz commented on audioipc/src/lib.rs:
```diff
@@ -61,11 +63,13 @@ pub type PlatformHandleType = std::os::windows::raw::HANDLE;
 
 unsafe impl Send for PlatformHandle {}
 
+const INVALID_HANDLE_VALUE: PlatformHandleType = -1isize as PlatformHandleType;
```
Good idea, will do.

kinetiknz commented on audioipc/src/lib.rs:
```diff
@@ -140,21 +141,26 @@ impl PlatformHandle {
     #[cfg(windows)]
-    pub unsafe fn into_file(self) -> std::fs::File {
-        std::fs::File::from_raw_handle(self.0)
+    pub unsafe fn into_file(&self) -> std::fs::File {
```
I'll add a check.

kinetiknz commented:
I've addressed the review comments, but I also reworked the patch.  Leaning on `Option.take` wasn't clear enough, I've dropped that and introduced an explicit flag to determine the ownership of the handle.

- Responsibility for closing the handle differs on Unix and Windows:
  - On Windows, handle is owned when deserialized, since the remote process already duplicated it into our process (and sent us the new handle value)
  - On Unix, handle is unowned when deserialized, since the remote process sends us its own handle values (which are just placeholders and ignored).  The handle becomes owned once it's read from the cmsghdr and updated in the ServerMessage via take_platform_handles.
- `new` and `into_raw` are now checked, which also means `into_file` is checked.

r? @ChunMinChang again please
